### ISSUE TITLE
Remove code that sets random prefix for PETSc options

### ISSUE THREAD

**Issue by abhyshr** [GridOPTICS/GridPACK#137]

**Comment by wperkins** [comment]
I agree that the option prefix problem is inconvenient. But, the prefix can't be simply removed without some strategy to replace what it does.  Do we remove the prefix completely, which means all instances of linear/nonlinear/ode/dae solvers in an application would be given the same options? Or, do we force a prefix to be provided in input?  Or, perhaps we need to rethink the exposure of PETSc options all together.  

IMO, GridPACK should not limit the mathematical methods available through PETSc. Users should be able to apply to their problem any linear/nonlinear/ODE/DAE/timestepping methods available.  

Also, IMO, if there are two linear solver instances, for example, in my GridPACK application, I want to be able to set the PETSc options for those independently.  

What we have now is what I could come up with to handle those two requirements.   There's probably a better way. I'm just not seeing it right now.

**Comment by abhyshr** [comment]
I am suggesting to remove the "random" prefix that GridPACK sets IF the user does not provide one. I can't exactly find where that piece of code is. You may be able to find it more easily than me. 

I don't quite understand how adding a prefix magically fixed the problem for Himanshu's case #136. My guess is some solver option was not being set correctly due to the prefix.

Any prefix stuff needs to be set by the user, not GridPACK, in my opinion.

**Comment by wperkins** [comment]
The decorating and insertion of options into the PETSc option database is handled by [`PETScConfigurable::p_processOptions()`](https://github.com/GridOPTICS/GridPACK/blob/3525b74d5a8f22a23127613ad30f688f29a31810/src/math/petsc/petsc_configurable.cpp#L143). 

The PETSc options database is created as PETSc is initialized with `gridpack::Environment`.  In #136, the same simulation was run over and over again using the same `gridpack::Environment` and PETSc options database.  The input indicates that there are 3  `LinearSolver` instances created by the application each time it is run.  Every creation of `LinearSolver` had a different PETSc option prefix generated, and the options decorated and inserted in the database.  This was done over and over and eventually overloaded the PETSc options database.  By manually setting the prefix, the names of the linear solver options do not change from run to run. The same options are repeatedly **replaced** in the options database. That way the database does not overload.

**Comment by bjpalmer** [comment]
It seems like the thing to do is to have some mechanism for setting a default, if you don't specify anything, something will happen and it will hopefully be fairly optimal. If you want to set things yourself, then you have that option. The defaults should be on an application basis so that powerflow uses a different default than dynamic simulation, etc.

I have a feeling that this could be handled fairly easily if we modified Configuration so that you could added fields by hand after reading in an external file.

**Comment by wperkins** [comment]
> It seems like the thing to do is to have some mechanism for setting a default, if you don't specify anything, something will happen and it will hopefully be fairly optimal. If you want to set things yourself, then you have that option. The defaults should be on an application basis so that powerflow uses a different default than dynamic simulation, etc.
> 
> I have a feeling that this could be handled fairly easily if we modified Configuration so that you could added fields by hand after reading in an external file.

I agree.  For `LinearSolver`, I would suggest that we have two preset configurations: "best" parallel direct (LU) and "best" parallel iterative.  These would be set up at build time.  The direct solver would be default and have a specific PETSc options prefix, like `direct_` and `iter_`. 

So, if SuperLU_dist is available, the best direct solver, which is currently set (if SuperLU_dist is available) using 
```
    <LinearSolver>
      <PETScOptions>
        -ksp_type richardson
        -pc_type lu
        -pc_factor_mat_solver_package superlu_dist
        -ksp_max_it 1
      </PETScOptions>
    </LinearSolver>
```
This would look like 
```
    <LinearSolver>
      <Type>Direct</Type>
    </LinearSolver>
```
Or, just
```
    <LinearSolver>
    </LinearSolver>
```
If one wants an iterative solver, I would suggest this configuration
```
      <LinearSolver>
        <SolutionTolerance>1.0E-08</SolutionTolerance>
        <MaxIterations>50</MaxIterations>
        <PETScOptions>
          -ksp_type bicg
          -pc_type bjacobi
          -sub_pc_type ilu -sub_pc_factor_levels 5 -sub_ksp_type preonly
        </PETScOptions>
      </LinearSolver>
```
which would be replaced with 
```
      <LinearSolver>
        <Type>Iterative</Type>
        <SolutionTolerance>1.0E-08</SolutionTolerance>
        <MaxIterations>50</MaxIterations>
      </LinearSolver>
```

There may be a larger set of `LinearSolver` configurations.  These seem to be the ones used most.

**Comment by wperkins** [comment]
Another thing to do would be having a consistent XML way to add `-*_monitor` and `-*_view` to solver options.  So, instead of doing this with a `LinearSolver`
```
      <LinearSolver>
        <Type>Iterative</Type>
        <SolutionTolerance>1.0E-08</SolutionTolerance>
        <MaxIterations>50</MaxIterations>
        <PETScOptions>
          -ksp_view
          -ksp_monitor
        </PETScOptions>
      </LinearSolver>
```
we would use something like
```
      <LinearSolver>
        <Type>Iterative</Type>
        <SolutionTolerance>1.0E-08</SolutionTolerance>
        <MaxIterations>50</MaxIterations>
        <Verbose>true</Verbose>
        <Summary>true</Summary>
      </LinearSolver>
```
The same options would be added to `NonlinearSolver` and `DAESolver`.

**Comment by wperkins** [comment]
I think this should be postponed.

**Comment by wperkins** [comment]
I still think this should be postponed.
